crate_universe/src/utils/starlark/serialize.rs

```diff
@@ -1,7 +1,86 @@
-use super::{ExportsFiles, Package};
-use serde::ser::{Serialize, SerializeStruct, SerializeTupleStruct, Serializer};
+use serde::ser::{SerializeSeq, SerializeStruct, SerializeTupleStruct, Serializer};
+use serde::Serialize;
 use serde_starlark::{FunctionCall, MULTILINE, ONELINE};
 
+use super::{
+    Data, ExportsFiles, Load, Package, RustBinary, RustLibrary, RustProcMacro, SelectList,
+};
+
+// For structs that contain #[serde(flatten)], a quirk of how Serde processes
+// that attribute is that they get serialized as a map, not struct. In Starlark
+// unlike in JSON, maps and structs are differently serialized, so we need to
+// help fill in the function name or else we'd get a Starlark map instead.
+pub fn rust_proc_macro<S>(rule: &RustProcMacro, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    FunctionCall::new("rust_proc_macro", rule).serialize(serializer)
+}
+
+pub fn rust_library<S>(rule: &RustLibrary, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    FunctionCall::new("rust_library", rule).serialize(serializer)
+}
+
+pub fn rust_binary<S>(rule: &RustBinary, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    FunctionCall::new("rust_binary", rule).serialize(serializer)
+}
+
+// Serialize an array with each element on its own line, even if there is just a
```

Comment:
I think we can remove this - our real goal is to be buildifier-compatible and buildifier accepts either, but happy to keep it also (or to remove it in a follow-up).

Reply:
This is worth keeping IMO. Note that the only case I use it in is inside `select`:

```bzl
    deps = select({
        "@rules_rust//rust/platform:aarch64-pc-windows-msvc": [
            "@cui__winapi-util-0.1.5//:winapi_util",  # cfg(windows)
        ],
        "@rules_rust//rust/platform:i686-pc-windows-msvc": [
            "@cui__winapi-util-0.1.5//:winapi_util",  # cfg(windows)
        ],
        "@rules_rust//rust/platform:x86_64-pc-windows-msvc": [
            "@cui__winapi-util-0.1.5//:winapi_util",  # cfg(windows)
        ],
        "//conditions:default": [],
    }),
```

compared to this which I find a little too compact to be easily readable, with the cfg comment way out of view:

```bzl
    deps = select({
        "@rules_rust//rust/platform:aarch64-pc-windows-msvc": ["@cui__winapi-util-0.1.5//:winapi_util"],  # cfg(windows)
        "@rules_rust//rust/platform:i686-pc-windows-msvc": ["@cui__winapi-util-0.1.5//:winapi_util"],  # cfg(windows)
        "@rules_rust//rust/platform:x86_64-pc-windows-msvc": ["@cui__winapi-util-0.1.5//:winapi_util"],  # cfg(windows)
        "//conditions:default": [],
    }),
```

Everywhere else I'm having arrays serialize using serde_starlark's default line wrapping, which is multiline for arrays &gt;1 element:

```bzl
    srcs = glob(["**/*.rs"]),
    compile_data = glob(
        include = ["**"],
        exclude = [
            "**/* *",
            "BUILD",
            "BUILD.bazel",
            "WORKSPACE",
            "WORKSPACE.bazel",
        ],
    ),
    tags = [
        "cargo-bazel",
        "manual",
        "noclippy",
        "norustfmt",
    ],
```